### PR TITLE
[DOC]: usage docs content guidelines

### PR DESCRIPTION
**Pull request by story645**
Per #26366, I am wary about a lot more work being put into revising the content of the user guide before we settle on what we think that content should read like.  I'm drafting some ideas here because I'd like us to adopt some sorta guidelines and also b/c I'd like discussion before I or anyone else puts the investment into getting our guide into this kinda form. The approach I describe here is inline with the approach we use in some tutorials, like [colormap tutorial](https://matplotlib.org/devdocs/users/explain/colors/colormap-manipulation.html#sphx-glr-users-explain-colors-colormap-manipulation-py) and [annotation tutorial](https://matplotlib.org/devdocs/users/explain/text/annotations.html). 

The reasons why I think we need content guidelines:
* a consistent approach to presenting the material is going to make the document far more cohesive, which makes it easier for a new user to follow the guide and for us to see what's missing. 
* much easier to intuit where a document is or should go if we have clear distinctions between "is it a guide, tutorial, or example?" -> @esibinga is doing this work for examples
* having guidelines we agree on before new content is written means less content to rewrite into the agreed upon form
* direction and structure if we'd want to hire a tech writer - and also for newer contributors who may feel overwhelmed by the whole thing
* streamlines reviews by having an agreed upon "this is what this should look like"

tldr proposal: https://github.com/matplotlib/matplotlib/pull/26389#issuecomment-1878141078

ETA: Pulling in from arguments elsewhere, style is not a matter of taste in this context because the disagreement we're having is on codifying which teaching methodologies to use in our most explanatory/educational set of docs.  Generally on the code side we ask that contributors follow best practice & that is considered a non-controversial review practice.  In the case of instructional documentation, there is also best practice on how to present new material to a novice audience. This PR is advocating for the adoption of some of these best practices as part of our content guidelines. Most of the  [PLOS 10 quick tips for teaching programming](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006023#pcbi.1006023.ref021) are transferable to teaching how to use a library.

ETA2: The Software Carpentry instructor training on managing [memory and cognitive load](https://carpentries.github.io/instructor-training/05-memory.html) discusses why the approach more or less described here is best practice, and we have very very high overlap w/ their audience, especially for introductory material.

### CODE REVIEW

**Comment by pukarlamichhane** [comment]
can I do this task

File: doc/devel/document.rst [hunk 1]
```diff
@@ -1002,6 +1002,79 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
 any ``*.rst`` files to a ``:toctree:``, either in the ``README.txt`` or in a
 manual ``index.rst``.
 
+
+.. _writing-user-guide:
+
+Write user guide explanations
+=============================
+
+The pages in the user guide are a mix of :ref:`ReST <writing-rest-pages>` and
```
Comment (timhoffm) [inline doc/devel/document.rst:1011]: Link "user guide" to the docs.

File: doc/devel/document.rst [hunk 2]
```diff
@@ -1002,6 +1002,79 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+:ref:`sphinx gallery <writing-examples-and-tutorials>` pages and should
+follow the respective formatting conventions. These pages are also intended to
```
Comment (timhoffm) [inline doc/devel/document.rst:1013]:
```suggestion
follow the respective formatting conventions. These pages are intended to
```
I'd say that is the only intention.

File: doc/devel/document.rst [hunk 3]
```diff
@@ -1002,6 +1002,79 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+explain how Matplotlib works to someone who is unfamiliar with the library. To
+maintain the consistency and cohesiveness of the user guide,  documentation
+that is contributed to this section should conform to the following guidelines:
+
+Purpose: Explain how to use Matplotlib
```
Comment (timhoffm) [inline doc/devel/document.rst:1018]:
```suggestion
Purpose: Explain how  Matplotlib works
```

"How tho use" sounds too much like "tutorial"

File: doc/devel/document.rst [hunk 4]
```diff
@@ -1002,6 +1002,79 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+known by the audience, but are used to explain what data ``Line2D`` accepts.
+
+
+Audience: new users
+-------------------
+
+The audience for this guide is getting introduced to using Matplotlib through
+this guide. Therefore content should be written with the assumption that the
+reader does not yet know what Matplotlib calls a given visualization task or
+how any task is accomplished in Matplotlib. Therefore each document should
```
Comment (timhoffm) [inline doc/devel/document.rst:1046]:
```suggestion
how any task is accomplished in Matplotlib. Therefore, each document should
```

File: doc/devel/document.rst [hunk 5]
```diff
@@ -1002,6 +1002,79 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+always first introduce or define the object/module/concept that it is discussing
```
Comment (timhoffm) [inline doc/devel/document.rst:1047]:
```suggestion
first introduce or define the object/module/concept that it is discussing
```
Maybe not needed?

File: doc/devel/document.rst [hunk 6]
```diff
@@ -1002,6 +1002,79 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
 any ``*.rst`` files to a ``:toctree:``, either in the ``README.txt`` or in a
 manual ``index.rst``.
 
+
+.. _writing-user-guide:
+
+Write user guide explanations
```
Comment (timhoffm) [inline doc/devel/document.rst:1008]:
```suggestion
Write user guide documentation
```
While the user guide is explanation-style, I prefer  the general therm "documentation " here. "User guide explanations" made me wonder if there are other parts of the user guide.

**Review by jklymak — COMMENTED** [review]
With regards to audience, on the whole I largely agree.  However, I disagree that beginners should be able to random access the user guide and necessarily understand any given subsection without having understood some basic concepts first.  Similarly, within a section, it should definitely begin straight forwardly and with strong motivation, but in further subsections we should be able to assume the reader has read the introductory material.  If every section is written assuming nothing else has been read, then the user guide will be infinitely long and unbearable to read for people who have actually read some of the intro material.

File: doc/devel/document.rst [hunk 7]
```diff
@@ -1045,6 +1045,104 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+The aim of the user guide is to teach the conceptual abstractions on which the
+Matplotlib API is developed so that users can understand how to fit the individual
+components of the library together. Therefore, content should be concept oriented
+rather than focused on specific tasks, e.g.
+
+  Lines in Matplotlib
+
+rather than *How do I make a squiggly yellow line?*
+
+Audience: new users
```
Comment (tacaswell) [inline doc/devel/document.rst:1073]: This should be "new to Matplotlib" not "new to Python", "new to numpy", or "new to programming".

If we find our selves explaining Python syntax or basic CS (like what objects are) we should instead be linking to CPython docs.

File: doc/devel/document.rst [hunk 8]
```diff
@@ -1045,6 +1045,82 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+
+
+Scope: Matplotlib objects and modules
+-------------------------------------
+
+Many concepts in Matplotlib assume a grounding in visualization, statistics, and
+other topics to understand how they work. These concepts should be
+contextualized using common terminilogy, but the focus should not stray from the
+Matplotlib topic, e.g. *`Line2D` objects take as input either pairwise coordinates
+(x,y) or y with an implicit x and assume that the input data is continuous.*
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/devel/document.rst]: Line doesn't "assume" anything.  Line draws something with a certain style.  Assumptions are for the person using it to draw the line, or the viewer.

File: doc/devel/document.rst [hunk 9]
```diff
@@ -1045,6 +1045,82 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+this section should conform to the following guidelines:
+
+Purpose: Explain how Matplotlib works
+-------------------------------------
+
+The aim of the user guide is to teach the concepts on which the Matplotlib API is
+developed so that users can understand how to fit the individual components of the
+library together. Therefore, content should be concept oriented rather than focused on
+specific tasks, e.g. *What is a `Line2D` artist?* rather than *How do I make a yellow
+squiggly line?*
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/devel/document.rst]: In my opinion the user guide should address "how to use Matplotlib to create data visualizations".  

"How to conceptualize Matplotlib", or "How does Matplotlib do x..." is hopefully inferred by example, and may be important to point out explicitly sometimes, but I don't think should be the focus of the User's Guide.  

I think the difference with the "Examples" section should be that there should be some narrative leading from basics to advanced, whereas the Examples can be curated stack overflow sized answers to "How do I do X" specific thing without too much concern for narrative.  That said, I think the "how to use Matplotlib to create data visualizations" should have many many examples.

File: doc/devel/document.rst [hunk 10]
```diff
@@ -1045,6 +1045,82 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+
+Teaching style: Chunking
+------------------------
+
+The material is introduced in small, usually one change or task at a time, chunks to
+keep focus on the specific line of code enabling the given task. The aim is that by
+reducing cognitive load, it will be easier for users to learn what each specific object
+does and how it behaves. The goal is that gaining a conceptual understanding of
+Matplotlib's API will make it easier for a user to navigate the rest of the
+documentation.
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/devel/document.rst]: Again, not a fan.  I think it's better to show the code, and its resultant plot, see the orange line, and then scan the code for the obvious place where the line was made orange.  People learn by repetition, and seeing the same code over and over and picking out the differences is better in my opinion than disambiguated code snippets in the middle of text.

**Comment by jklymak** [comment]
> Can you elaborate on this?

A lot of chunking or explanation for the initial parts of topics makes some sense.  However as we get deeper into a topic, the need for explanation and chunking should drop away as we assume more knowledge.  As a trivial example, we shouldn't have to explain what `fig, axs = plt.subplots()` means or does for every page, etc as we get deeper into the User Guide

File: doc/devel/document.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/devel/document.rst]: > We know that isn't true though b/c the feedback we got from the user survey was that folks don't feel like they can learn how to use Matplotlib from our docs 

I don't know where the results of that survey were tabulated.  However, I certainly agree that the docs when the survey was run were completely inadequate and confusingly organized.  I'm sure the current state of the docs is far from perfect, but it's also far from the state of the docs two years ago.

File: doc/devel/document.rst [hunk 11]
```diff
@@ -1045,6 +1045,104 @@ subdirectory, but  :file:`galleries/users_explain/artists` has a mix of
...
+
+Scope: Visualization using Matplotlib
+-------------------------------------
+
+Many concepts in Matplotlib assume a grounding in visualization, statistics, and
+other topics to understand how they work. These concepts should be
+contextualized using common terminilogy, but the focus should not stray from the
+Matplotlib topic, e.g.
+
+  ``Line2D`` objects take as input either pairwise coordinates (x,y) or y with an implicit
```
Comment (tacaswell) [inline doc/devel/document.rst:1101]: I find this example very confusing, "pairewise coordinates" usually means `[(x1, y1), (x2, y2), ...]` (which we will take and plot, but treat both columns as y.

File: doc/devel/content_guide.rst [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+Scope
+-----
+
+Many concepts in Matplotlib assume a grounding in visualization, statistics, Python
+programming, and other topics to understand how they work. These concepts should be
+contextualized using common terminology, but the focus for all documents should not
+stray from the Matplotlib topic, e.g.
+
+  ``Line2D`` objects take as input a list of pairwise coordinates (x,y) and generates
```
Comment (tacaswell) [inline doc/devel/content_guide.rst:107]: This is still at best extremely misleading https://github.com/matplotlib/matplotlib/blob/c287b9f985c3d2231b4c975f8ad7580ce8cbb8ce/lib/matplotlib/lines.py#L287-L308

as `Line2D` takes in an sequence for x and a sequence for y and they are zipped much deeper in the draw process.

File: doc/devel/content_guide.rst [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+organization of the documentation on using the library. Broadly, this documentations
+falls into one of these sections:
+
+:ref:`plot_types`
+    Summary of visualization techniques implemented as high-level API
+
+:ref:`users-guide-explain`
+    | Explanations of key concepts and core functionality,
+    | e.g. :ref:`what-is-a-backend`, :ref:`colors_def`, and  :ref:`event_handling`
+
```
Comment (jklymak) [inline doc/devel/content_guide.rst:17]: I'd suggest something much closer to:

```suggestion
:ref:`users-guide-explain`
    The users guide to Matplotlib, meant to be read in order with enough detail and examples 
    for users to be able to use Matplotlib effectively.  
```

I would consider the three examples you give to be considerable asides to the Users Guide, rather than representative of their main goal.  The previous users guide had things like that as the _only_ things in it, but that just made it a grab bag of disconnected subjects that no one would call a "User's Guide"

**Review by tacaswell — CHANGES_REQUESTED** [review]
There is no reference to https://diataxis.fr

**Comment by tacaswell** [comment]
I have several serious concerns

This makes no reference to https://diataxis.fr .  It is a well thought out, documented, and widely used system for organizing the docs.  I thought we had previously agreed we were going to start using.  We do not have the time or expertise to develop a new documentation system from scratch.  I think highlighting which of existing docs are clean example of their categories, but diataxis's categories should be the guiding principle of how the docs on our docs are written are framed.

I disagree about the level of the target audience of our docs.  Using software carpentry as benchmark, the floor of our target audience should be people who have completed Software Carpentry (not people who should take or are taking carpentry classes).

I am worried about framing our docs too closely through guidelines for developing lessons and in-person teaching of people who do not know how to program.  I think both the level is wrong (same as the previous concern) and documentation is consumed very differently with different goals.

I think the user guide should be roughly two parts.  The first which is a linear section that after you read you should understand all the parts of the `Figure` and how they interact with each other.  The second part which is deep dives explaining the more complicated APIs.  Both are "understand" but the first is at a system scale and second at a component granular scale.

@jklymak has been doing an extremely good job of organizing and filling in the first and most of what was in the user guide prior to Jody's work was the second.  I expect these docs to be validating and embracing that work rather than trying to do something different.

**Comment by jklymak** [comment]
> This makes no reference to https://diataxis.fr/ . It is a well thought out, documented, and widely used system for organizing the docs.

Is this the case?   Sorry to question a previous decision, but I am somewhat skeptical that the diataxis framework actually leads to useful organization of docs, at least for a library like Matplotlib.  Does anyone have a good example of it being implemented?  

> .... these docs to be validating and embracing that work

I'd also prefer we not re-define what should go in each section absent of the substantial work that has already been done.  However, I also agree that it is reasonable to write down what the plan is.

**Comment by jklymak** [comment]
I have a pretty different take on the slicing of things.  As noted a few times before, our previous organization was largely based around the fact folks wanted to write documentation using sphinx gallery.  Quite appropriately in my opinion, as it is easier to develop the visualizations that go along with the narrative if you can see them in quasi real time, rather than wait for a (partial) docs build.  Hence long-form write-ups ended up in "tutorials".  In my opinion, 90% of that material belonged in the "users guide". 

Second, I think we should be very clear that some of our main sections can and should have substantial overlap.  

Third, any organization needs to be flexible,  and we shouldn't get too hung up on _exactly_ where something goes.  I teach Physics, but choose your favourite subject, and pull two first year texts off the shelf, and the odds of them being in _exactly_ the same order, or having the same org tree is extremely small.  Particularly for online docs, I think being super fussy about the table of contents is counter productive.  

So given that, I'd divide things as 

"Users Guide" Narrative explanation for how to use Matplotlib, moving from general intorduction to specific subtopics and explanations, _heavily_ using examples.

"Examples" Short self contained documents that have one main subject.  If someone comes up with a new one, we should ask if it fits in the existing "Users Guide" or not, and maybe expand the scope to adding there is appropriate (or suggest a follow-up issue).  

"Tutorials": I think these should be cross-cutting such that they don't fit into the User Guide narrative, and they are too involved or too many topics for an Example.  I also think we should have very few of these.  

So let's take the recent discussion about histogram normalization (#27426).  I feel it would be a mistake to put this info in the "Tutorial" section.  Let's say a user found this tutorial, and then was trying to find it again a month later without remembering our TOC.  If we start putting specific subject docs like this under Tutorials, folks have three places to look: "Users Guide - hmm there is no discussion of _any_ plot methods - strange users guide", "Examples - ahhh, there are a bunch of histogram examples; its got to be one of these", "Tutorials - thats a weird place for this specific topic". Medium-term, I think it would be great if material from a few of the `hist` examples was combined to make a User Guide doc about how to use the `hist` method.  I don't feel that means we would have to get rid of the existing examples - we should just liberally cross-link.  

I think the above is really simple - everything basically belongs in the User Guide (eventually), supplemented by an extensive gallery with back links to the user guide, and with a few cool tutorials that combine a bunch of topics that would be too confusing in the middle of the User Guide.

**Comment by jklymak** [comment]
Repeating my objection: I strongly object to relegating the users guide to simply "learning/explanation" and think it needs to show you how to "push the buttons".  Ideally the users guide would be readable without reference to the examples or tutorials.  Having read it, the reader will come away understanding how to use Matplotlib and have some idea of the concepts that underpin how Matplotlib works.

**Comment by jklymak** [comment]
> This is sorta why my revised proposal is that the getting started guide does this & is in tutorial form. 

I feel the Getting Started Guide should be considerably shortened, rather than expanded.  When originally composed it was the only user guide that the User Guide section had, and picked up numerous digressions.  Much of the digressive material is now elsewhere in the User Guide.  Expanding the Getting Started would be going in the wrong direction.

**Comment by jklymak** [comment]
> What's the taxonomy here? I'm reading this and seeing the same spectrum we already have where:
> 
> 1. it's not clear why a document is placed where it is -> not for the reader nor for a ~new~ contributor

The "taxonomy" is trivial: 

If something is part of a self-contained subject:
  - it would _ideally_ be in the User Guide.
      - an accompanying example could _also_ be in the Examples.   
  - if there is no obvious TOC location for it in the User Guide, it could be parked in the appropriate section of Examples, and an issue raised that there is no proper location in the User Guide

If something is a long cross-cutting document (eg how to set up a dashboard) it would go in Tutorials.  

> 2. there is a lack of clear concrete guidance on what sort of subjects are appropriate and which aren't

Does the doc help someone make scientific visualizations in Matplotlib in general?  Then it belongs in the User Guide.  If it is some specialist visualization technique that doesn't particularly teach something that isn't already in the User Guide, then it should be an Example.  

> 3. docs within a page or section do not have a clear or readily apparent cohesive thread.

I'm not sure where you inferred this - The User Guide should be coherent.  

> 4. These are all the same document - for all intents and purposes the only difference is in length-which means that they inherently aren't tuned to serve different needs.

I don't think it particularly has to do with length.  Individual subsections of the User Guide need not be a particular length. Similarly I think it's fine if some Examples are long.  However, Examples should be on one topic, or they stretch out into being Tutorials.

**Comment by jklymak** [comment]
> what's missing from the rubric + tutorial getting started + explanation of concepts in terms of getting to this?

As before, your rubric says that "How to push the button" is under Examples, and not in the User Guide.  I strongly (extremely strongly) object to that.  "How to push the button" is the main point of the User Guide, explanation of concepts happens along the way by way of comprehensive buttton-pushing examples.

**Comment by jklymak** [comment]
> If the same document can be placed in user guide and examples or examples and tutorials, then that means those sections of the docs aren't serving distinct purposes. Fundamentally doc architecture works the same as API architecture -> clear well defined roles/scope/purpose - if they're not serving distinct purposes then it's unclear why they exist.

I think we are conflating "distinct purposes" and "distinct documents".   The "User Guide" needs examples, and the idea that those examples should not overlap with the "Examples" section is not viable, in my opinion. 

The difference between User Guide and Examples is not in content, but in how they are presented, accessed, and somewhat in how they are organized.  I'd think of it as akin to a text book, versus a study guide.  They probably overlap in a lot of their examples, but the text book has more verbiage, and the study guide has more worked examples.  

My comment about the same document going in the Users Guide or Examples was not meant to mean that _exactly_ the same document would go in both, verbatim.  Just that the contributor should be asked to consider putting it in the User Guide with an appropriate level of explanatory verbiage, if there is an appropriate place in the table of contents.  The more stripped down version may still be appropriate for the "Examples". If there is no place in the Users Guide table of contents, it can probably just go in Examples, but we should consider if the TOC needs fleshing out.

**Comment by jklymak** [comment]
> And now I'm thinking about disciplinary differences cause a CS textbook's sections would be topics akin to my proposal of functionality: variables, assignment, selection, repetition, etc -> the textbook introduces the concept and motivation for the concept, and then has a short worked example or two, then builds up to the next small worked example -> but big putting everything together happens in a case study at the end of the chapter or section that's much more akin to the tutorials.

That sounds great.  My  point is that they are all in the same chapter, and the user doesn't have to flip between whole different sections to find the concepts in use.  By analogy, I don't think the tutorial that "puts everything together" should be in a separate "Tutorial" section of the docs, any more than I think examples should be relegated to an "Examples" section.  

> Explanation: all about annotation in Matplotlib, with lots of very small very toy coding examples to keep focus on annotation -> in some ways this is just fleshed out API docs
https://matplotlib.org/devdocs/users/explain/text/annotations.html

This is far better than "talk about the buttons", with `Line2D` as an example.  

> Tutorials: really cute story of making donut plots, holds the readers hand through using annotations to highlight plot sections, annotation within context of larger task of building viz
https://matplotlib.org/devdocs/gallery/pie_and_polar_charts/pie_and_donut_labels.html
Also yes cheated and grabbed something from examples b/c one of the consequences of the current doc placement scheme is tutorials in the wrong places

I would still object to this - we don't need to squeeze the org into the diataxis framework. I'm sorry, but as a non-django user, the Django docs are pretty bad.  Apologies to the hard working folks who put in volunteer time to write them, but you'd have to come up with a far better example to convince me that diataxis is worth bending over backwards to fill out full sections for.  `pie_and_donut_labels` should stay in examples.  Why would we move it to Tutorials?  It's a single self contained example.  As stated before, I think we should have vanishingly few links in tutorials, and they should tackle pretty big cross cutting things.  Making donut graphs is not a big cross-cutting thing in my opinion, and I don't see how it would help discoverability or readability to have it off in Tutorials by itself where no one will ever find it, instead of in the "Pie and polar charts" section of the examples where it is easy to find.

**Comment by jklymak** [comment]
> Explain: What are the blocks? Why do we use 'em?

I wish you would stop calling the User Guide "Explain".  It is meant to be a User Guide, and "Why do we use 'em?" should be replaced by "How do we use them?".  I don't think. its a good idea, but if someone wants to write a "Theory of Matplotlib" section, they can go ahead, but lets not conflate the User Guide with that.  

> and moving the docs that are extend into their own section is a way to make it clear "hey, please don't let this overwhelm/intimidate you, by the time you need to know this, this will be accessible. 

If the goal is to have a TOC:

- Users Guide:
  - make:
     - figure
     - axes
     - line plot
     - pcolorplot
     - ...
  - customize:
     - figure
     - axes
     - line plot
     - pcolorplot
     - ...
  - extend:
     - figure, axes, line plot, pcolorplot, ...
- Tutorials
     - figure, axes, line plot, pcolorplot, ...
- Examples
     - figure, axes, line plot, pcolorplot, ...

Then I think this would be a big mistake.  My outline would rather be

- Users Guide:
   - figure:
     - make
     - customize
     - extend
   - axes:
     - make
     - customize
     - extend
   - plotting
     - lineplot
       - make, customize, extend
     - pcolorplot....
- Tutorials:
   - a couple of things
- Examples:
     Basically as they are now:
     - figure
     - axes
     - line plot
     - pcolorplot
     - ...
    
which is basically what we have now, though with some obvious gaps.  Ripping customization of all the methods out into their own sections doesn't make sense to me.  

> Yeah at some point before this goes back on the call in two weeks

I have no idea what call you are talking about, and I'll likely be on a ship.

**Comment by tacaswell** [comment]
The docs are much more random-access than linear read (a vast majority of people who land in our docs come in through search engines and land someplace deep within our docs).  While some sub-parts may make sense to be read linearly, overall they do not.  We should be careful about over-indexing either text book or lesson planning concepts onto the overall structure.

----

The point of diataxis is to sort out the purpose of a given document and then write for that use.  My attempt at translating this to what we currently have:

 - explain: the goal is to understand something.  In our case, that is the user guide where the first goal should be for the reader to end up with a (correct) mental model of how Matplotlib is designed.  There is no way to do that without examples and also teaching them how to use the library.  In here we don't need to cover _every_ formatter, but the idea that Axes have Axis that have both major and minor locators & formatters so that the user can go "oh, I want these ticks someplace else, I should look up `Locators`...".  This is exactly what I think you have been producing with your recent work @jklymak . I would also add a "deep dive" section that collects the explanations of some of our more complex APIs that include the why of the design which would smooth out some of the flow issues (e.g. no one should have to learn about event loop integration unless they really need to).   These are "breadth first" docs.
 - how-to: the goal is to quickly be able to do a thing without having to understand.  There are our examples as copy-paste-editable snippets.  You do not need to understand what a locator is to use an example that sets one if you are willing to accept some level of "magic" (which with computers we _always_ accept there is some level below which we do not need (want) to look).  The goal of these is to send the user off to solve their actual problem with a useful bit of code as quickly as possible without asking them to read a multi page essay waxing poetically about the details of our architecture or academic visualization theory.
 - tutorials: these are long-form prose documents that are some cross-cutting project that both shows the code to get some outcome and explains the choices that were made and why.  If you took a tutorial, deleted all of the text and only kept the final version you get get and example/how-to and if you took a how-to and bulked it up with incremental development of the code and prose about each step then you would get a tutorial.  In contrast to the user guide, these are "depth first" docs.
 - reference: the docstrings which have the exact spelling of the parameters in as terse of a way as we can achieve. 


The point of this sorting is not the _content_ but what the reader is trying to do at the time when they read it.   Are they trying to understand Matplotlib as a thing (user guide) or are they just trying to get the plot they need on the screen ASAP (examples/reference)?  Do they want to they want general explanations (user guide) or worked complex examples (tutorials)?


-----


I agree not putting code in the user guide is a non-starter, but I don't think anyone is proposing that.

**Comment by jklymak** [comment]
> what I was proposing was more like:

That is a couple of examples, and I don't know how you would implement it in general.  

I'm fine with @tacaswell's [characterization of the docs](https://github.com/matplotlib/matplotlib/pull/26389#issuecomment-1881545586) :

> The point of diataxis is to sort out the purpose of a given document and then write for that use. My attempt at translating this to what we currently have...

and would prefer we continued to work in this direction.

**Comment by jklymak** [comment]
> Frankly I don't understand how what I proposed wasn't already working in this direction. In this discussion and in the PR as it currently stands, the language isn't quite the same (and yes I'll revise) but is aimed at getting to the same thing:

>> explain: the goal is to understand something. In our case, that is the user guide where the first goal should be for the reader to end up with a (correct) mental model of how Matplotlib is designed.

Because that sounds like the User Guide is a discourse on how the library was designed, _not_ on how to _use_ it.  It would be like if I got the User Guide to a fridge and all its "first goal" was to explain how refrigeration works.  Most people don't care - they want the User Guide to tell them how to _use_ the fridge.  

Users Guide: the goal is to understand how to use Matplotlib.

**Comment by jklymak** [comment]
> Is the hangup in how I'm using the word "works"?
>
> ETA: "Explain how to use" -> "Explain how it works" was a PR suggestion in https://github.com/matplotlib/matplotlib/pull/26389#discussion_r1429321822 - me and @timhoffm were using them interchangebly

Yes that is the hang up.  A user only needs to know how it "works" to get things done.  I disagree with @timhoffm's comment that there is anything wrong with this sounding like a tutorial, and strongly prefer how you originally wrote it.  

> The second sentence explicitly says that by "works=functionality", like a manual that explains how the fridge works talks about the icemaker and the freezer and has an exploded view of the fridge. The rest of that section is like "identify each component" like how a fridge manual says "and the icemaker is here"

I agree.   How to "find" the ice maker is a great thing for the Users _Guide_.  How to "use" the ice maker is a good thing for the _Users_ Guide.  Contrary to @tacaswell, I would put "How does the ice maker work?" at the end of the list, as something that is perhaps helpful for a user to understand but not 100% necessary.  But a full explanation of the Carnot cycle would probably belong elsewhere.

File: doc/devel/document_content.rst [hunk 14]
```diff
@@ -0,0 +1,457 @@
+.. _documenting-content:
+
+******************
+Content guidelines
+******************
```
Comment (timhoffm) [inline doc/devel/document_content.rst:5]:
```suggestion
********************************
Documentation content guidelines
********************************
```

**Comment by timhoffm** [comment]
@story645 I have read the majority of this now (but not all details of the *User guide* section).

- The fundamental idea of explicitly writing down the purpose and how to write the different parts of our documentation is great. - And I believe basically all core devs will agree with this.
- I think, you're doing too much in this PR, which is why the PR is not moving forward well: 1. It's too much in volume - a reviewer needs quite some time to dig into this (this is also why I myself pushed looking into this forward again and again). 2. It's too much in complexity, in particular the "User guide" section is quite a construct, and by itself a possible source of longer debate.
- Apart from the review aspect: Can we make this more compact? The document feels quite long, and the longer it is the less people will read it. I propose as a general guiding question: With how little content and conceptual overhead can we reasonably cover the topic? - I'm particularly thinking of the "User guide" here, but also e.g. of the introduced term "modality" for the type of documentation.

To move forward, I propose to split this up and in a first step cut off after the summary (probably simplest to 1:1 copy that part to a new PR and start discussing there). The detailed descriptions of the individual sections can be handled later.

**Comment by story645** [comment]
Just a note that the current status on this is that this has functionally become an issue on improving doc guidance, with the following complete:

- plot types: #30084
- gallery documentation:  #28527
- api guidance: #30085

And the next steps are:
1. package summary as preamble/heading to docs guide 
2. revive #27769
3. open PR focusing on usage guide guidance
